typed-transitions/src/Protocol/Driver.hs

```diff
@@ -0,0 +1,88 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE RankNTypes #-}
+
+module Protocol.Driver where
+
+import Data.Text (Text)
+
+import Protocol.Channel
+import Protocol.Core
+import Protocol.Codec
+
+data Result t where
+  Normal     :: t -> Result t
+  -- | Unexpected data was given. This includes the case of an EOF.
+  Unexpected :: Text -> Result t
+  deriving (Show)
+
+-- | Drive a 'Peer' using a 'Duplex', by way of a 'Codec' which describes
+-- the relationship between the concrete representation understood by the
+-- 'Duplex', and the typed transitions understood by the 'Peer'.
+--
+-- A failure to decode arises as an 'Unexpected :: Result t'.
+useCodecWithDuplex
+ :: forall m concreteSend concreteRecv p tr status init end t .
+    ( Monad m )
+ => Duplex m m concreteSend concreteRecv
+ -> Codec m concreteSend concreteRecv tr init
+ -> Peer p tr (status init) end m t
+ -> m (Result t)
+useCodecWithDuplex = go Nothing
```

Comment:
Just to be sure that I understand this correctly.  There are three streams here: peer & decoder which match and streams of some encoded values given by `Duplex` (e.g. `ByteString`).  The peer type and its matching decoder orchestrate the read of these values from the channel, e.g. from types we will know that we are reading a transition from the current state to some continuation state, the codec gives as a way to decode them (usually a partial decoder) which will digest bytes from the channel.

Adding this kind of high level explanation might be useful for the future.

Reply:
I would also add a link to this part in `Protocol.Codec` since it gives the interpretation of types defined there.

Reply:
I'll add some documentation.